// @ciscospark/plugin-people/dist/people-batcher.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _sparkCore = require('@ciscospark/spark-core');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @class
 */
var PersonUUIDRequestBatcher = _sparkCore.Batcher.extend({
  namespace: 'People',

  /**
   * @param {HttpResponseObject} res
   * @returns {Promise}
   */
  handleHttpSuccess: function handleHttpSuccess(res) {
    var _this = this;

    return _promise2.default.all(res.body.items.map(function (personResponse) {
      return _this.handleItemSuccess(personResponse.id, personResponse);
    }));
  },
  handleItemFailure: function handleItemFailure(email, response) {
    return this.getDeferredForResponse(email).then(function (defer) {
      defer.reject(response);
    });
  },


  /**
   * @param {string} email
   * @param {Object} response
   * @returns {Promise}
   */
  handleItemSuccess: function handleItemSuccess(email, response) {
    return this.getDeferredForResponse(email).then(function (defer) {
      defer.resolve(response);
    });
  },


  /**
   * @param {string} uuid
   * @returns {Promise<string>}
   */
  fingerprintRequest: function fingerprintRequest(uuid) {
    return _promise2.default.resolve(this.spark.people.inferPersonIdFromUuid(uuid));
  },


  /**
   * @param {string} uuid
   * @returns {Promise<string>}
   */
  fingerprintResponse: function fingerprintResponse(uuid) {
    return _promise2.default.resolve(uuid);
  },
  prepareRequest: function prepareRequest(ids) {
    return _promise2.default.resolve(ids.join());
  },


  /**
   * @param {Object} ids
   * @returns {Promise<HttpResponseObject>}
   */
  submitHttpRequest: function submitHttpRequest(ids) {
    return this.spark.request({
      service: 'hydra',
      resource: 'people/?id=' + ids
    });
  }
}); /**!
     *
     * Copyright (c) 2015-2017 Cisco Systems, Inc. See LICENSE file.
     * @private
     */

exports.default = PersonUUIDRequestBatcher;
